compiler/catala_utils/messages.ml

```diff
@@ -2,188 +2,232 @@
-
-let debug_format (format : ('a, Format.formatter, unit) format) =
-  if !Cli.debug_flag then
-    Format.printf ("%a@[<hov>" ^^ format ^^ "@]@.") debug_marker ()
-  else Format.ifprintf Format.std_formatter format
-
-let error_format (format : ('a, Format.formatter, unit) format) =
-  Format.print_flush ();
-  (* Flushes previous warnings *)
-  Format.printf ("%a" ^^ format ^^ "\n%!") error_marker ()
+type content_type = Error | Warning | Debug | Log | Result
 
-let warning_format format =
-  if !Cli.disable_warnings_flag then Format.ifprintf Format.std_formatter format
-  else Format.printf ("%a" ^^ format ^^ "\n%!") warning_marker ()
+let get_ppf = function
+  | Result -> Lazy.force std_ppf
+  | Debug when not !Cli.debug_flag -> Lazy.force ignore_ppf
+  | Warning when !Cli.disable_warnings_flag -> Lazy.force ignore_ppf
+  | Error | Log | Debug | Warning -> Lazy.force err_ppf
+
+(**{3 Markers}*)
+
+let print_time_marker =
+  let time : float ref = ref (Unix.gettimeofday ()) in
+  fun ppf () ->
+    let new_time = Unix.gettimeofday () in
+    let old_time = !time in
+    time := new_time;
+    let delta = (new_time -. old_time) *. 1000. in
+    if delta > 50. then
+      Format.fprintf ppf "@{<bold;black>[TIME] %.0fms@}@\n" delta
+
+let pp_marker target ppf () =
+  let open Ocolor_types in
+  let tags, str =
+    match target with
+    | Debug -> [Bold; Fg (C4 magenta)], "[DEBUG]"
+    | Error -> [Bold; Fg (C4 red)], "[ERROR]"
+    | Warning -> [Bold; Fg (C4 yellow)], "[WARNING]"
+    | Result -> [Bold; Fg (C4 green)], "[RESULT]"
+    | Log -> [Bold; Fg (C4 black)], "[LOG]"
+  in
+  if target = Debug then print_time_marker ppf ();
+  Format.pp_open_stag ppf (Ocolor_format.Ocolor_styles_tag tags);
+  Format.pp_print_string ppf str;
+  Format.pp_close_stag ppf ()
 
-let result_format format =
-  Format.printf ("%a" ^^ format ^^ "\n%!") result_marker ()
+(**{2 Printers}*)
 
-let log_format format =
-  Format.printf ("%a@[<hov>" ^^ format ^^ "@]@.") log_marker ()
+(** Prints the argument after the correct marker and to the correct channel *)
+let format target format =
+  let ppf = get_ppf target in
+  pp_marker target ppf ();
+  Format.pp_print_space ppf ();
+  Format.pp_open_hovbox ppf 0;
+  Format.kfprintf
+    (fun ppf ->
+      Format.pp_close_box ppf ();
+      Format.pp_print_newline ppf ())
+    ppf format
 
 (** {1 Message content} *)
 
 module Content = struct
-  type position = { message : string option; position : Pos.t }
-  type t = { message : string; positions : position list }
+  type message = Format.formatter -> unit
+  type position = { pos_message : message option; pos : Pos.t }
+  type t = { message : message; positions : position list }
+
+  let of_message (message : message) : t = { message; positions = [] }
 
-  let of_message (s : string) : t = { message = s; positions = [] }
+  let of_string (s : string) : t =
+    { message = (fun ppf -> Format.pp_print_string ppf s); positions = [] }
```

Comment:
I think the codebase of the compiler uses `fmt` instead of `ppf` as the go-to variable names for formatters.